### ISSUE TITLE
Full screen distortion issues in Megaman Legends 2

### ISSUE THREAD

**Issue by Red-gehog** [stenzek/duckstation#1113]
Version: 0.1-2363-gfcbd6187 (RetroArch core)

Game: Megaman Legends 2 (USA)

System info

- Windows 10 64 bits (X86) 

This problem occurs at native resolution, but there are also problems with this effect when pgxp is enabled or texture filtering, both at increased resolution. If the internal res is increased without pgxp and texture filtering, there will only be a few '' ghosts '', as in the corner of the floor ahead. So I assumed that the problem is not the increase in res.

Both happen in opengl and vulkan, I couldn't test it in d3d11. Note that without pgxp and without texture filtering the effect is equivalent to internal res, but with the features already mentioned above, the effect seems to lose res.

![Mega Man Legends 2 (USA)-201128-053016](https://user-images.githubusercontent.com/73415642/100498085-7ffdb980-313e-11eb-97fe-4796a7dd9c8f.png)
Software (all hacks and enhancements are disabled)

![Mega Man Legends 2 (USA)-201128-053544](https://user-images.githubusercontent.com/73415642/100498111-c3582800-313e-11eb-984f-80b711592ed9.png)
Hardware, 4x internal res, no texture filtering, no pgxp

![Mega Man Legends 2 (USA)-201128-053342](https://user-images.githubusercontent.com/73415642/100498371-82611300-3140-11eb-812d-96d2e434c1ab.png)
Hardware, 4x internal res, with texture filtering,  pgxp correction on

![Mega Man Legends 2 (USA)-201128-053605](https://user-images.githubusercontent.com/73415642/100498411-bb998300-3140-11eb-87e5-1bff43c72f95.png)
Hardware, 4x internal res, only with texture filtering, no  pgxp correction

The problems are the same if you repeat the same situations of internal res, pgxp and texture filtering as above in sofware mode. With jinc and bilinear also happens the same.

Edit: In all modes for the the 1x res, only ghosts will appears.

**Comment by stenzek** [comment]
Could you please upload a memory card where this bug occurs?

**Comment by foxsevent** [comment]
> Could you please upload a memory card where this bug occurs?

I can help with that if it's all the same.
[MegaMan Legends 2_2.zip](https://github.com/stenzek/duckstation/files/5618432/MegaMan.Legends.2_2.zip)

After you load the save turn left and enter the door ahead, afterwards take the elevator in the centre of the map down, after which it's just a straight line to the door and the room in the first post.

Cheers,

**Comment by Red-gehog** [comment]
Thanks foxsevent! These are the coordinates. This is also the same effect applied to the screen when it happens in the underwater parts.

Here is the memory card you ordered:

[Mega Man Legends 2 (USA)_1.zip](https://github.com/stenzek/duckstation/files/5620229/Mega.Man.Legends.2.USA._1.zip)

**Comment by stenzek** [comment]
Hmm, not sure if I'm doing something wrong - the haze effect appears to be working here?

https://www.duckstation.org/upload/Mega%20Man%20Legends%202%20(USA)%202020-12-02%2000-23-08.mp4

Obviously there will be a difference between resolution scales with these kinds of effect, but I'm not sure what's broken at 1x?

**Comment by Red-gehog** [comment]
The problem happens when the resolution is increased using pgxp and texture filtering, the distortion effect that simulates heat (and that also simulates water in other places in the game) seems to lose resolution.

**Comment by Red-gehog** [comment]
From what I saw, you switched the driver to vulkan, increased the scale of res. In the screenshots I posted, see that in the second image where the internal res is increased, but without pgxp and texture filtering, everything is at high res. In the third image, the res is also increased, but with pgxp enabled, and texture filtering enabled, see that the distortion effect over the screen has lost resolution.

**Comment by Red-gehog** [comment]
You need to enable pgxp and increased texturing filtering to see the effect at low res.

**Comment by stenzek** [comment]
Visual effects like this will break with increased resolution, this is expected. Using texture filtering will also break things.

This isn't really a bug unless it's broken at native resolution, which I'm still not seeing?

**Comment by Red-gehog** [comment]
So this is what you said, filtering with internal upscaling is breaking the effect, and pgxp also breaks it along with internal upscaling. It is a pity.

Is there a way to prevent filtering or pgxp from breaking the screen distortion effect? Something like the option to ignore 2d sprites and 2d polygons from filtering (opaque, transparent or both) similar to the Beetle PSX HW?

**Comment by stenzek** [comment]
So I looked into how this was rendering - it draws a bunch of one-pixel high polygons, sampling with slightly offset texture coordinates from the same framebuffer it's writing to. Kinda insane.

Why it breaks with PGXP - UVs are clamped to the polygon bounds when PGXP or texture filtering is on. This range is multiplied by the resolution scale, but because it's a range of 1, it only allows sampling from the original (downscaled) texel.

I've fixed this in https://github.com/stenzek/duckstation/commit/59ac365b52304432d1c68b8a5c00b5412f27c5e8 by extending the range to include the additional pixels which would fall inbetween the original texel. Only affects framebuffer affects with PGXP on, so the chances of breakage are slim, and the couple of other games with framebuffer effects which I tried seem fine.
